banman/store.go

```diff
@@ -0,0 +1,217 @@
+// backed by a database.
+func newBanStore(db walletdb.DB) (*banStore, error) {
+	s := &banStore{db: db}
+
+	// We'll ensure the expected buckets are created upon initialization.
+	err := walletdb.Update(db, func(tx walletdb.ReadWriteTx) error {
+		banStore, err := tx.CreateTopLevelBucket(banStoreBucket)
+		if err != nil {
+			return err
+		}
+		_, err = banStore.CreateBucketIfNotExists(banBucket)
+		if err != nil {
+			return err
+		}
+		_, err = banStore.CreateBucketIfNotExists(reasonBucket)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// BanPeer creates a ban record for the peer within the BanStore for the given
+// duration. A reason can also be provided to note why the peer is being banned.
+// The record will exist until a call to BanStatus is made after the ban
+// expiration.
+func (s *banStore) BanPeer(addr string, reason Reason, duration time.Duration) error {
+	// We'll extract the host so we can consider it without taking into
+	// account the target port.
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return err
+	}
+
+	return walletdb.Update(s.db, func(tx walletdb.ReadWriteTx) error {
+		banStore := tx.ReadWriteBucket(banStoreBucket)
+		if banStore == nil {
+			return ErrCorruptedBanStore
+		}
+		banIndex := banStore.NestedReadWriteBucket(banBucket)
+		if banIndex == nil {
+			return ErrCorruptedBanStore
+		}
+		reasonIndex := banStore.NestedReadWriteBucket(reasonBucket)
+		if banIndex == nil {
+			return ErrCorruptedBanStore
+		}
+		return addBannedPeer(
+			banIndex, reasonIndex, host, reason, duration,
+		)
+	})
+}
+
+// addBannedPeer adds an entry to the BanStore for the given peer.
+func addBannedPeer(banIndex, reasonIndex walletdb.ReadWriteBucket, host string,
+	reason Reason, duration time.Duration) error {
+
+	k := []byte(host)
+	var v [8]byte
+	banExpiration := time.Now().Add(duration)
+	byteOrder.PutUint64(v[:], uint64(banExpiration.Unix()))
+
+	if err := banIndex.Put(k, v[:]); err != nil {
+		return err
+	}
+	return reasonIndex.Put(k, []byte{byte(reason)})
+}
+
+// BanStatus returns the ban status for a given peer.
+func (s *banStore) BanStatus(addr string) (BanStatus, error) {
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return BanStatus{}, err
+	}
+
+	var banStatus BanStatus
+	err = walletdb.Update(s.db, func(tx walletdb.ReadWriteTx) error {
+		banStore := tx.ReadWriteBucket(banStoreBucket)
+		if banStore == nil {
+			return ErrCorruptedBanStore
+		}
+		banIndex := banStore.NestedReadWriteBucket(banBucket)
+		if banIndex == nil {
+			return ErrCorruptedBanStore
+		}
+		reasonIndex := banStore.NestedReadWriteBucket(reasonBucket)
+		if banIndex == nil {
+			return ErrCorruptedBanStore
+		}
+
+		k := []byte(host)
+		status := fetchBanStatus(banIndex, reasonIndex, k)
+
+		// If the peer's ban duration has expired, we can remove its
+		// entry from the store.
+		if !time.Now().Before(status.Expiration) {
```

Comment:
Style nit: I usually find usage of `After` more readable, so: 
```go
if time.Now().After(status.Expiration)
```

Reply:
That wouldn't handle the case of `time.Now() == status.Expiration`, though you could say it's not crucial.